### PR TITLE
CASMPET-6173: Updates to prerequisites.sh to aide cert-manager changes

### PR DESCRIPTION
**Pull request by mitchty**
# Description

# Checklist

- [ ] If I added any command snippets, the steps they belong to follow the prompt conventions (see [example][1]).
- [ ] If I added a new directory, I also updated `.github/CODEOWNERS` with the corresponding team in [Cray-HPE][2].
- [ ] My commits or Pull-Request Title contain my JIRA information, or I do not have a JIRA.

 
[1]: https://github.com/Cray-HPE/docs-csm/blob/main/introduction/documentation_conventions.md#using-prompts
[2]: https://github.com/Cray-HPE/teams

### CODE REVIEW

File: upgrade/scripts/upgrade/prerequisites.sh [hunk 1]
```diff
@@ -880,6 +880,43 @@ else
     echo "====> ${state_name} has been completed" | tee -a "${LOG_FILE}"
 fi
 
+state_name="UPGRADE_CERTMANAGER_CHART"
+state_recorded=$(is_state_recorded "${state_name}" "$(hostname)")
+if [[ ${state_recorded} == "0" && $(hostname) == "${PRIMARY_NODE}" ]]; then
+    echo "====> ${state_name} ..." | tee -a "${LOG_FILE}"
+    {
+        # Only remove these charts if installed
+        if helm list -n cert-manager --filter cray-certmanager | grep cray-certmanager > /dev/null 2>&1; then
```
Comment (bklei) [inline upgrade/scripts/upgrade/prerequisites.sh:889]: Seems like this grep could get tricked by the cray-certmanager-issuers chart?

```
ncn-m001:~/bklein # helm list -n cert-manager --filter cray-certmanager
NAME                    	NAMESPACE   	REVISION	UPDATED                                	STATUS  	CHART                         	APP VERSION
cray-certmanager        	cert-manager	3       	2023-03-02 16:39:09.968339672 +0000 UTC	deployed	cray-certmanager-0.6.0        	0.14.1
cray-certmanager-issuers	cert-manager	3       	2023-03-02 16:39:19.103486396 +0000 UTC	deployed	cray-certmanager-issuers-0.6.2	0.14.1
```

File: upgrade/scripts/upgrade/prerequisites.sh [hunk 2]
```diff
@@ -880,6 +880,43 @@ else
     echo "====> ${state_name} has been completed" | tee -a "${LOG_FILE}"
 fi
 
+state_name="UPGRADE_CERTMANAGER_CHART"
+state_recorded=$(is_state_recorded "${state_name}" "$(hostname)")
+if [[ ${state_recorded} == "0" && $(hostname) == "${PRIMARY_NODE}" ]]; then
+    echo "====> ${state_name} ..." | tee -a "${LOG_FILE}"
+    {
+        # Only remove these charts if installed
+        if helm list -n cert-manager --filter cray-certmanager | grep 'cray-certmanager$' > /dev/null 2>&1; then
```
Comment (johren-hpe) [inline upgrade/scripts/upgrade/prerequisites.sh:889]: Needs a `-q` if you want to have the $ at the end of the grep.   
Does the --filter do anything?

File: upgrade/scripts/upgrade/prerequisites.sh [hunk 3]
```diff
@@ -880,6 +880,43 @@ else
...
+            helm uninstall -n cray-certmanager cray-certmanager
+        fi
+        if helm list -n cray-certmanager-init --filter cray-certmanager-init | grep cray-certmanager-init > /dev/null 2>&1; then
```
Comment (johren-hpe) [inline upgrade/scripts/upgrade/prerequisites.sh:892]: Does the --filter do the same thing as the grep?

File: upgrade/scripts/upgrade/prerequisites.sh [hunk 4]
```diff
@@ -880,6 +880,43 @@ else
...
+apiVersion: manifests/v1beta1
+metadata:
+  name: cray-certmanager-images-tmp-manifest
+spec:
+  charts:
+  -
+EOF
+
+        yq r "${CSM_MANIFESTS_DIR}/platform.yaml" 'spec.charts.(name==cray-drydock)' | sed 's/^/    /' >> "${tmp_manifest}"
+        yq r "${CSM_MANIFESTS_DIR}/platform.yaml" 'spec.charts.(name==cray-certmanager)' | sed 's/^/    /' >> "${tmp_manifest}"
```
Comment (johren-hpe) [inline upgrade/scripts/upgrade/prerequisites.sh:907]: The hyphens aren't ending up right.   This is what I get:
```
apiVersion: manifests/v1beta1
metadata:
  name: cray-certmanager-images-tmp-manifest
spec:
  charts:
  -
    name: cray-drydock
    source: csm-algol60
    version: 2.18.0
    namespace: loftsman
    name: cray-certmanager
    source: csm-algol60
    version: 0.7.2
    namespace: cert-manager
```

File: upgrade/scripts/upgrade/prerequisites.sh (hunk 3 above)
Comment (mitchty) [reply upgrade/scripts/upgrade/prerequisites.sh:892]: Yes but helm list doesn't change the return code so the greps really there just to make $? 0 or 1

**Review by bo-quan — CHANGES_REQUESTED** [review]
See inline namespace corrections

**Review by bo-quan — APPROVED** [review]

File: upgrade/scripts/upgrade/prerequisites.sh [hunk 5]
```diff
@@ -880,6 +880,62 @@ else
...
+        start=$(date +%s)
+        lim=1
+        until ! kubectl get namespace "${cmns}" > /dev/null 2>&1; do
+          now=$(date +%s)
+          if [ "$((now-start))" -ge 300 ]; then
+            printf "fatal: namespace %s likely requires manual intervention after waiting for removal for at least 5 minutes, details:\n" "${cmns}" >&2
+            kubectl get namespace "${cmns}" -o yaml
+            exit 1
+          fi
+          lim="$((lim*2))"
```
Comment (bo-quan) [inline upgrade/scripts/upgrade/prerequisites.sh:921]: The exponentially increasing wait could end up waiting for more than 8 minutes in the unhappy path: 1, 2, 4, 8, 16, 32, 64, 128, 256sec, then exit. Checking the condition every 5 or 10 seconds may be a better solution.

File: upgrade/scripts/upgrade/prerequisites.sh (hunk 4 above)
Comment (rustydb) [reply upgrade/scripts/upgrade/prerequisites.sh:926]: @mitchty can you use `yq4` instead of `yq`? It will help us switch from yq version 3.

**Pull request merged**